### PR TITLE
Create `off` method to de-register from upstream events

### PR DESCRIPTION
**Pull request by jd-carroll**

### CODE REVIEW

**Comment by jd-carroll** [comment]
I don't see why this is failing..
```
...
onScriptLoad@file:///home/travis/build/dmvaldman/samsara/test/vendor/require.js:30:186: undefined is not an object (evaluating 'this.upstream[i].off')
```

Shouldn't all upstream objects have an `off`?

File: samsara/events/EventHandler.js [hunk 1]
```diff
@@ -115,6 +115,26 @@ define(function(require, exports, module) {
...
+     * @method off
+     * @param type {String}           Event channel name
+     * @param [handler] {Function}    Handler
+     */
+    EventHandler.prototype.off = function off(type, handler) {
+        var empty = EventEmitter.prototype.off.apply(this, arguments);
+        if (empty && this.upstreamListeners[type]) {
+            var oldUpstreamListener = this.upstreamListeners[type];
+            delete this.upstreamListeners[type];
+            for (var i = 0; this.upstream.length; i++) {
```
Comment (dmvaldman) [inline samsara/events/EventHandler.js]: `i < this.upstream.length` :-)

**Comment by dmvaldman** [comment]
Found the problem. Thought I was the only one who messed up for loop expressions, but now there's two of us ;-)

Put up some more tests, and this PR (with the fixes) passes them. So looking forward to merging in the fixed version.

File: samsara/events/EventHandler.js (hunk 1 above)
Comment (jd-carroll) [inline samsara/events/EventHandler.js]: lol...

**Comment by jd-carroll** [comment]
Should be all set.

The best is when I write syntactically correct loops from Java / .Net / JS in the wrong language...

**Pull request merged**